File: main/contrib/Windows API Code Pack 1.1/source/Samples/ExtendedLinguisticServices/Transliterator/VB/ScrollbarTextBox.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.

Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Diagnostics

Namespace Transliterator
	Public Class ScrollbarTextBox
		Inherits TextBox
		Private Const WM_HSCROLL As Integer = &H114
		Private Const WM_VSCROLL As Integer = &H115

		Public Event OnHorizontalScroll As ScrollEventHandler
		Public Event OnVerticalScroll As ScrollEventHandler

        Private Shared ScrollEventTypeArray() As ScrollEventType = {ScrollEventType.SmallDecrement, ScrollEventType.SmallIncrement, ScrollEventType.LargeDecrement, ScrollEventType.LargeIncrement, ScrollEventType.ThumbPosition, ScrollEventType.ThumbTrack, ScrollEventType.First, ScrollEventType.Last, ScrollEventType.EndScroll}

		Private Function GetEventType(ByVal wParam As UInteger) As ScrollEventType
            If wParam < ScrollEventTypeArray.Length Then
                Return ScrollEventTypeArray(CInt(wParam))
            Else
                Return ScrollEventType.EndScroll
            End If
		End Function

		Protected Overrides Sub WndProc(ByRef m As Message)
			MyBase.WndProc(m)

			If m.Msg = WM_HSCROLL Then
				If OnHorizontalScrollEvent IsNot Nothing Then
					Dim wParam As UInteger = CUInt(m.WParam.ToInt32())
                    RaiseEvent OnHorizontalScroll(Me, New ScrollEventArgs(GetEventType(CUInt(wParam And &HFFFF)), CInt(Fix(wParam >> 16))))
				End If
			ElseIf m.Msg = WM_VSCROLL Then
				If OnVerticalScrollEvent IsNot Nothing Then
					Dim wParam As UInteger = CUInt(m.WParam.ToInt32())
                    RaiseEvent OnVerticalScroll(Me, New ScrollEventArgs(GetEventType(CUInt(wParam And &HFFFF)), CInt(Fix(wParam >> 16))))
				End If
			End If
		End Sub
	End Class
End Namespace

File: main/contrib/Windows API Code Pack 1.1/source/Samples/ExtendedLinguisticServices/Transliterator/CS/ScrollbarTextBox.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Diagnostics;

namespace Transliterator
{
    public class ScrollbarTextBox : TextBox
    {
        private const int WM_HSCROLL = 0x114;
        private const int WM_VSCROLL = 0x115;

        public event ScrollEventHandler OnHorizontalScroll = null;
        public event ScrollEventHandler OnVerticalScroll = null;

        private static ScrollEventType[] scrollEventType = new ScrollEventType[] 
        {   ScrollEventType.SmallDecrement,
            ScrollEventType.SmallIncrement,
            ScrollEventType.LargeDecrement,
            ScrollEventType.LargeIncrement,
            ScrollEventType.ThumbPosition,
            ScrollEventType.ThumbTrack,
            ScrollEventType.First,
            ScrollEventType.Last,
            ScrollEventType.EndScroll
        };

        private ScrollEventType GetEventType(uint wParam)
        {
            if (wParam < scrollEventType.Length)
            {
                return scrollEventType[wParam];
            }
            else
            {
                return ScrollEventType.EndScroll;
            }
        }

        protected override void WndProc(ref Message m)
        {
            base.WndProc(ref m);

            if (m.Msg == WM_HSCROLL)
            {
                if (OnHorizontalScroll != null)
                {
                    uint wParam = (uint)m.WParam.ToInt32();
                    OnHorizontalScroll(this, new ScrollEventArgs(GetEventType(wParam & 0xffff), (int)(wParam >> 16)));
                }
            }
            else if (m.Msg == WM_VSCROLL)
            {
                if (OnVerticalScroll != null)
                {
                    uint wParam = (uint)m.WParam.ToInt32();
                    OnVerticalScroll(this, new ScrollEventArgs(GetEventType(wParam & 0xffff), (int)(wParam >> 16)));
                }
            }
        }
    }
}
